senders/msteams/msteams.go

```diff
@@ -0,0 +1,207 @@
+package msteams
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/moira-alert/moira"
+	"github.com/russross/blackfriday/v2"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// Sender implements moira sender interface via MS Teams
+type Sender struct {
+	frontURI string
+	logger   moira.Logger
+	location *time.Location
+	client   *http.Client
+}
+
+// Init initialises settings required for full functionality
+func (sender *Sender) Init(senderSettings map[string]string, logger moira.Logger, location *time.Location, dateTimeFormat string) error {
+	sender.logger = logger
+	sender.location = location
+	sender.frontURI = senderSettings["front_uri"]
+	sender.client = &http.Client{
+		Timeout: time.Duration(30) * time.Second,
+	}
+	return nil
+}
+
+// SendEvents implements Sender interface Send
+func (sender *Sender) SendEvents(events moira.NotificationEvents, contact moira.ContactData, trigger moira.TriggerData, plot []byte, throttled bool) error {
+
+	err := sender.isValidWebhookURL(contact.Value)
+	if err != nil {
+		return fmt.Errorf("invalid msteams webhook url: %s", err.Error())
+	}
+
+	request, err := sender.buildRequest(events, contact, trigger, plot, throttled)
+
+	if request != nil {
+		defer request.Body.Close()
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to build request: %s", err.Error())
+	}
+
+	response, err := sender.client.Do(request)
+	if response != nil {
+		defer response.Body.Close()
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to perform request: %s", err.Error())
+	}
```

Comment:
I'd also rewrite this to be more in line with Go http module documentation:
```
if err != nil {
  return fmt.Errorf("failed to perform request: %s", err.Error())
}
defer response.Body.Close()
```

I see one more problem here. As per Go http module docs, `A non-2xx status code doesn't cause an error`. So this code will emit an error log on timeout or context cancellation, but no error on 500 response code from MS Teams API. I think we should check for `response.StatusCode`.

And another problem. As per docs, `If the Body is not both read to EOF and closed, the Client's underlying RoundTripper (typically Transport) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request`. So we have to read the body to EOF, not just close.

See https://golang.org/pkg/net/http/#Client.Do